app/master/slave.py

```diff
@@ -47,6 +50,10 @@ def mark_as_idle(self):
 
         self.current_build_id = None
 
+        if self._is_shutdown:
```

Comment:
I noticed in ClusterSlave.kill() there is still a todo comment to shutdown more gracefully. Correct me if I am wrong, but I feel this change addresses that todo. Maybe we should remove it, or open an issue for it.